vscode-dotnet-runtime-library/src/EventStream/OutputChannelObserver.ts

```diff
@@ -54,6 +55,11 @@ export class OutputChannelObserver implements IEventStreamObserver {
                     this.stopDownloadIndicator();
                 }
                 break;
+            case EventType.DotnetAcquisitionSuccessEvent:
+                if (event instanceof DotnetExistingPathResolutionCompleted) {
+                    this.outputChannel.append(`Using configured .NET path: ${ (event as DotnetExistingPathResolutionCompleted).resolvedPath }\n`);
```

Comment:
This is definitely a hammer. I might make this log appear more like a warning stating that there's a variable overriding dynamic acquisition. The reason why that's important is because users applications may fail because we're pointing our dotnet acquisition to some random .NET and not verifying it's the correct kind. Also, by overriding the acquired dotnet you opt-out of servicing which is also awful.

All that being said. This configuration option might need some more thought. The .NET acquisition extension is built to handle multiple extensions using it simultaneously. Imagine you have 3 different extensions using this extension and there's an override for dotnet? You're then assuming the one dotnet on the box can handle very single VSCode extension.

Reply:
Oh and to be clear I like how you allow multiple versions but you're then making the assumption that users know which version of dotnet to force. Also, having an "all" is dangerous for reasons above 😄

Reply:
> I might make this log appear more like a warning stating that there's a variable overriding dynamic acquisition

Good idea, will do. 

> This configuration option might need some more thought.

Yeah, I was struggling to come up with a clear format for the configuration that balances usability and configuration for multiple extensions/versions. 
One other option is to specify the extension ID (ex. `msazurermtools.azurerm-vscode-tools`) instead of the version. Then the `all` option doesn't make sense anyways, so we can remove that and users don't have to know what version of .NET the extensions use. The downside here is that they won't know which version to manually install if they don't already have it on the machine.

Reply:
> The downside here is that they won't know which version to manually install if they don't already have it on the machine.

ya that's a big concern I have as well. To me instead of having this configuration option it makes me feel like there should be "installable" workloads that install directly into the VSCode extension directory for the .NET core acquisition extension. That way if users have trouble dynamically acquiring pieces you can point them directly to an offline install. With that approach you'd maintain the ability to service the "offline installed" dotnet's. Just a thought though.

Reply:
I don't think anyone on our team would be thrilled about a new method of installation, it's hard to maintain.

If the main concern is that the user doesn't know which version of .NET to manually install, it shouldn't be that hard to expose that information to the user. It's already present in the output channel and we could add it to the error popup as well, something along the lines of `.NET Installation Failed (3.1): ...`?

Reply:
> I don't think anyone on our team would be thrilled about a new method of installation, it's hard to maintain.

Totally understand.



> If the main concern is that the user doesn't know which version of .NET to manually install, it shouldn't be that hard to expose that information to the user. It's already present in the output channel and we could add it to the error popup as well, something along the lines of `.NET Installation Failed (3.1): ...`?

You could but you also don't want users mistaking network blips as "I need to go and install something else offline".

Reply:
> You could but you also don't want users mistaking network blips as "I need to go and install something else offline".

I don't necessarily see this as being any more encouraging to the user to try a work around than the existing behavior. This would just be adding the version number to the existing message, I wouldn't expect the average user to see that and jump to trying to manually install, unless the problem is recurrent. 

@wli3 do you have any thought here?

Reply:
> I wouldn't expect the average user to see that and jump to trying to manually install, unless the problem is recurrent.

That's fair. I only bring it up because O# who does similar acquisition logic has yet to make that leap to include the info in a pop-up and instead has README documentation in the GitHub repo on how to offline install their bits.

Reply:
Yeah, I was thinking about how we could keep it in docs instead of the popup, but the problem there is that we would have to maintain a list of extensions who depend on us and which version they require instead of getting it dynamically based on what they request. It seems like it would be problematic to maintain such a list.

Reply:
Also including the version in the error popup doesn't directly tell the user to manually install. In fact, it could make sense to do this in general- consider the case when two extensions are acquiring different versions and one succeeds and the other fails.

Reply:
I am thinking about 2 ways.

1. instead of keeping track of which extension want which local build. We could override it by runtime version. And use the local runtime when the runtime version the extension asked is compatible.

2. I am now sure this using local runtime feature need to be generalized. We could add/keep this functionality in ARM extension to avoid regression. Due to the complexity mentioned above, each extension could decide if to use local runtime first and then run the acquisition extension.

Reply:
@StephenWeatherford per the discussion above we are considering cutting this feature (the built in configuration option to point us to an existing .NET install). What is your motivation for wanting it here instead of in the ARM extension?

Reply:
The motivation is that the dotnet extension is supposed to be solving the installation problem for everyone.  But there are multiple known scenarios where your installation doesn't work (no PowerShell, group policies, old .NET, etc.).  Ideally the installation scripts should handle this correctly, or there should be options for using something other than PowerShell).  I can't imagine we're the only ones facing that problem.

So all I care about is that the dotnet installation should either work in all configurations or the should guide users on how to fix it, manually install, or work around it without each individual extension having to code for the scenario. Otherwise it defeats the whole purpose.  And I'm fine with offline docs as long as you point to them where the user can easily find them.

Reply:
@StephenWeatherford Yes, I see where you're coming from and we are definitely aware of the issues with the install scripts, there is an active effort to improve them now. I agree that if there's a clean way to do this we should, the concern is that anything we can do on our end will be more confusing to the user than helpful. 

Just to lay things out more clearly for my own understanding, I see the main decision as which format we want to user to specify the configuration in: 
1. [runtime version, path] or 
2. [extension, path]. 

The complexity is that the user may not know which runtime a given extension needs.

@wli3 to clarify, your suggestion would be to have the user specify the version and location of the runtime (option 1), with the note that we would use a specified runtime as long as it is compatible (ie the version the user specifies does not have to exactly match the version the extension is requesting as long as they are compatible)? Do you have any thoughts on including the runtime version the extension is requesting in the error popup to give the user more guidance on which version to install?

Reply:
> The motivation is that the dotnet extension is supposed to be solving the installation problem for everyone. But there are multiple known scenarios where your installation doesn't work (no PowerShell, group policies, old .NET, etc.)

I do like this long term vision. However, this is a very complex problem space. A large portion of the team's effort is spent on acquisition (via VS, docker, install script). However all of them have restrictions on different situations. It will be an on going effort to resolve these problems. And we do need to rank them with other bugs we have. cc @marcpopMSFT 

> Do you have any thoughts on including the runtime version the extension is requesting in the error popup to give the user more guidance on which version to install?

That should also be fine. If we cannot download them. It is still a path lead to solution of this problem

Reply:
@wli3 I definitely agree that we need to rank new feature requests among the rest of our team's workload. In this case, it looks like we're coming to the conclusion that the right thing to do here is to allow the user to specify the version and the .NET path, which is what I've already implemented in this PR. It would take a few minutes to add the runtime version number to error popups, which mitigates the risk of users not knowing which runtime to install. Unless you see any other complexities here or someone disagrees that this is the best path forward, I think it's worthwhile for me to make that quick addition and then merge these changes, since I've already implemented this.

Reply:
I see this extension as trying to help people get .net installed in most cases and Iike the fallback option this PR adds of them installing separately and specifying a version.  As Will notes, we have limited time and can't solve every install case in every tool that install .net core so just relying on the fallback and allowing for some failures in this extension make sense to me.